### PR TITLE
Replace Actor Channel's pubsub with grpc

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR removes the usage of the redis actor pubsub channel with the grpc based pubsub.

It should also be more efficient because we should no longer be broadcasting every actor change to every node. Instead nodes should only receive updates from relevant actors. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/pubsub/publisher.h [hunk 1]
```diff
@@ -33,36 +33,68 @@ using SubscriberID = UniqueID;
 
 namespace pub_internal {
 
+class SubscriptionIndexInterface {
```
Comment (wuisawesome) [inline src/ray/pubsub/publisher.h:36]: These publisher changes could probably be pulled out into a separate PR, but they're pretty self-contained (makes the `SubscriptionIndex` follow the same pattern as the subscriber's `SubscriberChannel` to "properly" support templating.

**Comment by wuisawesome** [comment]
This isn't ready to merge yet, but most of the logic is here, and I think it would be good to get and some initial thoughts even before all the bugs are fixed up.

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 2]
```diff
@@ -260,89 +261,139 @@ Status ServiceBasedActorInfoAccessor::AsyncSubscribe(
...
-        fetch_done(status);
-      }
+  if (grpc_based_pubsub_) {
+    auto &subscriber = client_impl_->GetGcsSubscriber();
+    auto sub_message = std::make_unique<rpc::SubMessage>();
+    rpc::Address gcs_address;
+    gcs_address.set_ip_address(client_impl_->GetGcsServerAddress().first);
+    gcs_address.set_port(client_impl_->GetGcsServerAddress().second);
+    auto subscription_callback = [actor_id, subscribe](const rpc::PubMessage &message) {
+      RAY_LOG(ERROR) << "Got a subscription!";
```
Comment (ericl) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:271]: Stray error log

**Comment by rkooo567** [comment]
Btw, can you add an unit test that checks if different type of IDs are working properly? (just one unit test to check if ObjectID and NodeID both are working with the subscriber / publisher)

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 3]
```diff
@@ -260,89 +261,139 @@ Status ServiceBasedActorInfoAccessor::AsyncSubscribe(
...
+      subscribe(id, message.actor_table_data());
+    };
+    auto failure_callback = [done](const std::string &message) {
+      done(Status::UnknownError(message));
+    };
+    subscriber.Subscribe(std::move(sub_message), rpc::ChannelType::GCS_ACTOR_CHANNEL,
+                         gcs_address, actor_id.Binary(), subscription_callback,
+                         failure_callback);
+    return Status::OK();
+  } else {
```
Comment (ericl) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:285]: Can you un-indent the else block? That would avoid such a huge diff below which presumably hasn't changed. There's an early return above so it's not needed.

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 4]
```diff
@@ -260,89 +261,139 @@ Status ServiceBasedActorInfoAccessor::AsyncSubscribe(
...
+        rpc::ChannelType::GCS_ACTOR_CHANNEL, gcs_address, actor_id.Binary());
+    if (success) {
+      return Status::OK();
+    } else {
+      std::stringstream msg;
+      msg << "Couldn't unsubscribe to notifications for actor: " << actor_id << ".";
+      return Status::UnknownError(msg.str());
+    }
+
+  } else {
```
Comment (ericl) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:340]: Please un-indent

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 5]
```diff
@@ -260,89 +261,139 @@ Status ServiceBasedActorInfoAccessor::AsyncSubscribe(
...
-    RAY_LOG(INFO) << "Finished fetching all actor information from gcs server after gcs "
-                     "server or pub-sub server is restarted.";
-  };
+  if (grpc_based_pubsub_) {
+    // TODO (Alex): We'll need to handle this case for HA GCS. (Ideally we
+    // can handle this in the pubsub layer and restore state by resending the
+    // command batch).
+    RAY_LOG(FATAL)
+        << "GCS seems to have died. We've lost our subscription to actor changes.";
+  } else {
```
Comment (ericl) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:359]: Please un-indent

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc [hunk 6]
```diff
@@ -74,6 +75,9 @@ Status ServiceBasedGcsClient::Connect(instrumented_io_context &io_service) {
         RayConfig::instance().gcs_service_connect_retries()))
         << "Failed to get gcs server address when init gcs client.";
   }
+  subscriber_ = std::make_unique<pubsub::Subscriber>(
+      subscriber_id_, /*TODO*/ 10,
```
Comment (ericl) [inline src/ray/gcs/gcs_client/service_based_gcs_client.cc:79]: Put the real variable name here instead of TODO

File: src/ray/gcs/gcs_client/service_based_gcs_client.h [hunk 7]
```diff
@@ -64,12 +67,16 @@ class RAY_EXPORT ServiceBasedGcsClient : public GcsClient {
   /// Reconnect to GCS RPC server.
   void ReconnectGcsServer();
 
-  std::shared_ptr<RedisClient> redis_client_;
+  const UniqueID subscriber_id_;
```
Comment (ericl) [inline src/ray/gcs/gcs_client/service_based_gcs_client.h:70]: Comment

File: src/ray/gcs/gcs_server/gcs_server_main.cc [hunk 8]
```diff
@@ -62,7 +62,8 @@ int main(int argc, char *argv[]) {
     RAY_CHECK(status.ok()) << "Failed to init redis gcs client as " << status;
 
     // Init storage.
-    auto storage = std::make_shared<ray::gcs::RedisGcsTableStorage>(redis_client);
+    auto storage =
+        std::make_shared<ray::gcs::RedisGcsTableStorage>(redis_client, nullptr);
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_server_main.cc:66]: Please add /*var=*/ for nullptr, here and elsewhere.

File: src/ray/gcs/gcs_server/gcs_table_storage.cc [hunk 9]
```diff
@@ -152,5 +152,26 @@ template class GcsTableWithJobId<ObjectID, ObjectLocationInfo>;
...
+        publish_change)
+    : GcsTableWithJobId(store_client, publish_change) {
+  table_name_ = TablePrefix_Name(TablePrefix::ACTOR);
+}
+
+Status GcsActorTable::Put(const ActorID &key, const ActorTableData &value,
+                          const StatusCallback &callback) {
+  return GcsTableWithJobId::Put(key, value, [this, key, value, callback](Status status) {
+    callback(status);
+    if (status.ok() && publish_change_) {
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_table_storage.cc:168]: Instead of allowing publish_change_ to be nullptr, can you make it always a valid callback?

Also, please rename it to something like publish_via_grpc_

File: src/ray/gcs/gcs_server/gcs_table_storage.h [hunk 10]
```diff
@@ -294,6 +311,33 @@ class GcsInternalConfigTable : public GcsTable<UniqueID, StoredConfig> {
...
+  explicit GcsTableStorage(
+      std::shared_ptr<StoreClient> store_client,
+      std::function<void(const rpc::ChannelType channel_type,
+                         rpc::PubMessage &pub_message, const std::string &key_id_binary)>
+          publish_change)
+      : store_client_(store_client),
+        publish_change_(publish_change),
+        job_table_(std::make_unique<GcsJobTable>(store_client_)),
+        // TODO before merging, feature flag this.
+        actor_table_(std::make_unique<GcsActorTable>(store_client_, publish_change)),
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_table_storage.h:323]: Isn't it already flagged?

File: src/ray/pubsub/publisher.h (hunk 1 above)
Comment (rkooo567) [reply src/ray/pubsub/publisher.h:36]: Makes sense. This is to put different templated index to the same dict right?

**Review by rkooo567 — COMMENTED** [review]
General direction lgtm

File: src/ray/pubsub/subscriber.cc [hunk 11]
```diff
@@ -49,6 +44,7 @@ bool SubscriberChannel<KeyIdType>::Unsubscribe(const rpc::Address &publisher_add
   cum_unsubscribe_requests_++;
   const auto publisher_id = PublisherID::FromBinary(publisher_address.worker_id());
   const auto key_id = KeyIdType::FromBinary(key_id_binary);
+  RAY_LOG(ERROR) << "Attempting to unsubscribe from: " << publisher_id << "\t" << key_id;
```
Comment (ericl) [inline src/ray/pubsub/subscriber.cc:47]: Stray error

File: src/ray/pubsub/subscriber.cc [hunk 12]
```diff
@@ -84,6 +80,9 @@ SubscriptionCallback SubscriberChannel<KeyIdType>::GetCallbackForPubMessage(
   auto subscription_it = subscription_map_.find(publisher_id);
   // If there's no more subscription, do nothing.
   if (subscription_it == subscription_map_.end()) {
+    for (const auto &pair : subscription_map_) {
+      RAY_LOG(ERROR) << "\t" << pair.first;
```
Comment (ericl) [inline src/ray/pubsub/subscriber.cc:84]: Stray error

File: src/ray/gcs/gcs_server/gcs_publisher_manager.cc [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+  for (const auto &command : request.commands()) {
+    if (command.has_unsubscribe_message()) {
+      grpc_publisher_.UnregisterSubscription(command.channel_type(), subscriber_id,
+                                             command.key_id());
+    } else if (command.has_subscribe_message()) {
+      grpc_publisher_.RegisterSubscription(command.channel_type(), subscriber_id,
+                                           command.key_id());
+    } else {
+      RAY_LOG(ERROR) << "Invalid command has received, "
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_publisher_manager.cc:75]: Please change this to RAY_LOG(FATAL) and remove the GitHub message.

File: src/ray/gcs/gcs_server/gcs_publisher_manager.cc [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+                     << static_cast<int>(command.command_message_one_of_case())
+                     << ". If you see this message, please "
+                        "report to Ray "
+                        "Github.";
+      std::stringstream stream;
+      stream << "Invalid pubsub command received: " << request.DebugString();
+      status = Status::Invalid(stream.str());
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_publisher_manager.cc:82]: Don't need this after FATAL

File: src/ray/gcs/gcs_server/gcs_publisher_manager.h [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+#include "ray/common/asio/periodical_runner.h"
+#include "ray/pubsub/publisher.h"
+#include "ray/rpc/gcs_server/gcs_rpc_server.h"
+
+namespace ray {
+namespace gcs {
+
+/// This implementation class of `PublisherHandler`
+/// All non-handler public methods are threadsafe.
+class GcsPublisherManager : public rpc::PublisherHandler {
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_publisher_manager.h:26]: Why not call it *Handler since it's just a rpc handler?

File: src/ray/gcs/gcs_server/gcs_publisher_manager.h [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+                               rpc::SendReplyCallback callback);
+  void HandlePubsubCommandBatch(const rpc::PubsubCommandBatchRequest &request,
+                                rpc::PubsubCommandBatchReply *reply,
+                                rpc::SendReplyCallback callback);
+
+ private:
+  /// The thread + io service the publisher runs on.
+  std::unique_ptr<std::thread> publisher_thread_;
+  instrumented_io_context &publisher_service_;
+  PeriodicalRunner periodical_runner_;
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_publisher_manager.h:59]: We don't seem to ever use this member

File: src/ray/gcs/gcs_server/gcs_publisher_manager.h (hunk 15 above)
Comment (wuisawesome) [reply src/ray/gcs/gcs_server/gcs_publisher_manager.h:26]: I think the convention is Manager since I suppose it manages the lifecycle of the pubsub module? For example, `GcsHeartbeatManager` also doesn't really manage anything.
